gateway/src/apicast/policy/keycloak_role_check/keycloak_role_check.lua

```diff
@@ -0,0 +1,195 @@
+--       ]
+--     }
+--   }
+-- }
+--
+-- And you need to specify the "client_roles" in this policy as follows:
+--
+-- "client_roles": [
+--   { "name": "<client_role_A>", "client": "<client_A>" },
+--   { "name": "<client_role_B>", "client": "<client_A>" },
+--   { "name": "<client_role_A>", "client": "<client_B>" },
+--   { "name": "<client_role_B>", "client": "<client_B>" }
+-- ]
+
+local policy = require('apicast.policy')
+local _M = policy.new('Keycloak Role Check Policy')
+
+local ipairs = ipairs
+local MappingRule = require('apicast.mapping_rule')
+local TemplateString = require('apicast.template_string')
+local default_type = 'plain'
+
+local new = _M.new
+
+local function build_templates(scopes)
+  for _, scope in ipairs(scopes) do
+
+    if scope.realm_roles then
+      for _, realm_role in ipairs(scope.realm_roles) do
+        realm_role.template_string = TemplateString.new(
+          realm_role.name, realm_role.name_type or default_type)
+      end
+    end
+
+    if scope.client_roles then
+      for _, client_role in ipairs(scope.client_roles) do
+        client_role.name_template_string = TemplateString.new(
+          client_role.name, client_role.name_type or default_type)
+
+        client_role.client_template_string = TemplateString.new(
+          client_role.client, client_role.client_type or default_type)
+      end
+    end
+
+    scope.resource_template_string = TemplateString.new(
+      scope.resource, scope.resource_type or default_type)
+
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.type = config.type or "whitelist"
+  self.scopes = config.scopes or {}
+
+  build_templates(self.scopes)
+
+  return self
+end
+
+local function check_roles_in_token(role, roles_in_token)
+  for _, role_in_token in ipairs(roles_in_token) do
+    if role == role_in_token then return true end
+  end
+
+  return false
+end
+
+local function match_realm_roles(scope, context)
+  if not scope.realm_roles then return true end
+
+  for _, role in ipairs(scope.realm_roles) do
+    if not context.jwt.realm_access then
+      return false
+    end
+
+    local name = role.template_string:render(context)
+
+    if not check_roles_in_token(name, context.jwt.realm_access.roles) then
+      return false
+    end
+  end
+
+  return true
+end
+
+local function match_client_roles(scope, context)
+  if not scope.client_roles then return true end
+
+  for _, role in ipairs(scope.client_roles) do
+    if not context.jwt.resource_access then
+      return false
+    end
+
+    local client = role.client_template_string:render(context)
+    local client_in_token = context.jwt.resource_access[client]
+
+    if not client_in_token then
+      ngx.log(ngx.DEBUG, "Client '", client, "' was not found in the access token.")
+      return false
+    end
+
+    local name = role.name_template_string:render(context)
+
+    if not check_roles_in_token(name, client_in_token.roles) then
```

Comment:
Will `client_in_token.roles` always be != nil?

The function `check_roles_in_token()` expects a table and it will crash with nil. Maybe we should use `client_in_token.roles or {}` to be sure?

Reply:
According to Keycloak specification, `client_in_token.roles` always be != nil.
However, just to be on the safe side, I changed this.